### PR TITLE
Add healthchecks [RHELDST-23402]

### PR DESCRIPTION
**Pull request by drepelov**
This commit extends the `/status` endpoint to also check the status of workers, redis, celery beat and connection to gitlab and pulp.

### CODE REVIEW

File: ubi_manifest/app/api.py [hunk 1]
```diff
@@ -10,12 +13,70 @@
...
+        else:
+            days = time_from_last_heartbeat.days
+            hours = time_from_last_heartbeat.seconds // 3600
+            minutes = (time_from_last_heartbeat.seconds - hours * 3600) // 60
+            celery_beat_status = (
+                f"Someting is wrong, last heartbeat task ran "
+                f"{days} days, {hours} hours and {minutes} minutes ago."
+            )
+    else:
+        celery_beat_status = "No heartbeat task ran yet. Wait a minute."
```
Comment (rbikar) [inline ubi_manifest/app/api.py:55]: I would probably  rather see something like `n/a` instead of long message. Or you can introduce a bit more complicated output - that will have short status field (ok, n/a, failed), and longer msg field.

```
Celery beat status:
   status: OK
   msg: Celery beat operable
```

File: ubi_manifest/app/api.py [hunk 2]
```diff
@@ -10,12 +13,70 @@
...
+
+    # Check connection to GitLab
+    gitlab_resp = requests.get(
+        "https://gitlab.cee.redhat.com/-/health", timeout=REQUEST_TIMEOUT
```
Comment (rbikar) [inline ubi_manifest/app/api.py:59]: 1) This shouldn't be hardcoded but taken from config
2) If we use loading from file instead of GitLab - check should be skipped.

File: ubi_manifest/app/api.py [hunk 3]
```diff
@@ -10,12 +13,70 @@
...
 
+REQUEST_TIMEOUT = 20
+
 router = APIRouter(prefix="/api/v1")
 
 
 @router.get("/status")
-def status() -> dict[str, str]:
-    return {"status": "OK"}
+def status() -> dict[str, Any]:
```
Comment (rbikar) [inline ubi_manifest/app/api.py:22]: Can you please add documentation for this endpoint - at least how expected response looks and maybe response_model? For caller it's difficult to guess what to format of json.
E.g. `Workers status` has a bit complicated structure and in case we use multiple workers, it is currently not obvious for caller how the output looks.

File: ubi_manifest/worker/tasks/repo_monitor.py
Comment (rbikar) [inline ubi_manifest/worker/tasks/repo_monitor.py:1]: Can you please move those changes to different PR? Thx.

File: ubi_manifest/app/api.py [hunk 4]
```diff
@@ -10,12 +13,70 @@
...
+                f"{days} days, {hours} hours and {minutes} minutes ago."
+            )
+    else:
+        celery_beat_status = "No heartbeat task ran yet. Wait a minute."
+
+    # Check connection to GitLab
+    gitlab_resp = requests.get(
+        "https://gitlab.cee.redhat.com/-/health", timeout=REQUEST_TIMEOUT
+    )
+    gitlab_resp.raise_for_status()
```
Comment (rbikar) [inline ubi_manifest/app/api.py:61]: I think we should catch all exceptions raised from this call and set the status as `n/a` or `failed`.

File: ubi_manifest/app/api.py [hunk 5]
```diff
@@ -10,12 +13,70 @@
...
+    gitlab_status = gitlab_resp.text
+
+    # Check connection to Pulp
+    pulp_resp = requests.get(
+        f"{app.conf.pulp_url}/pulp/api/v2/status", timeout=REQUEST_TIMEOUT
+    )
+    pulp_resp.raise_for_status()
```
Comment (rbikar) [inline ubi_manifest/app/api.py:68]: I think we should catch all exceptions raised from this call and set the status as `n/a` or `failed`.

File: ubi_manifest/app/api.py [hunk 6]
```diff
@@ -10,12 +13,70 @@
...
+    pulp_status = pulp_resp.json()
```
Comment (rbikar) [inline ubi_manifest/app/api.py:69]: This status has pretty complicated output, maybe checking status code would be enough.

File: ubi_manifest/app/api.py [hunk 7]
```diff
@@ -10,12 +13,70 @@
...
+            )
+    else:
+        celery_beat_status = "No heartbeat task ran yet. Wait a minute."
+
+    # Check connection to GitLab
+    gitlab_resp = requests.get(
+        "https://gitlab.cee.redhat.com/-/health", timeout=REQUEST_TIMEOUT
+    )
+    gitlab_resp.raise_for_status()
+    gitlab_status = gitlab_resp.text
```
Comment (rbikar) [inline ubi_manifest/app/api.py:62]: Not sure how complicated this output is, but maybe just checking status code for 200 would be enough.

**Review by rbikar — CHANGES_REQUESTED** [review]

File: ubi_manifest/app/api.py [hunk 8]
```diff
@@ -1,21 +1,120 @@
...
+                        "server_status": "OK",
+                        "workers_status": {
+                            "availability": {"worker01": {"ok": "pong"}},
+                            "stats": {"worker01": {"some": "stats"}},
+                            "registered_tasks": {"worker01": ["some_task"]},
+                            "active_tasks": {"worker01": []},
+                            "scheduled_tasks": {"worker01": []},
+                        },
+                        "redis_status": {"Connected": True},
+                        "celery_beat_status": "OK",
```
Comment (rbikar) [inline ubi_manifest/app/api.py:43]: the structure of  `celery_beat_status` is bit complicated now, can you please update it here?

File: ubi_manifest/app/api.py [hunk 9]
```diff
@@ -1,21 +1,120 @@
...
+            "status": "n/a",
+            "msg": "No heartbeat task ran yet. Wait a minute.",
+        }
+
+    # Gitlab is used only if the content configs are loaded from Gitlab.
+    # In case the configs are loaded from directory, connection to Gitlab
+    # is not needed, therefore not checked.
+    gitlab_base_url = get_gitlab_base_url(app.conf.content_config)
+    if gitlab_base_url:
+        gitlab_resp = requests.get(
```
Comment (rbikar) [inline ubi_manifest/app/api.py:96]: Check the comment below.

File: ubi_manifest/app/api.py [hunk 10]
```diff
@@ -1,21 +1,120 @@
...
+            f"{gitlab_base_url}/-/health", timeout=REQUEST_TIMEOUT
+        )
+        gitlab_status = gitlab_resp.reason
+    else:
+        gitlab_status = "not needed"
+
+    # Check connection to Pulp
+    pulp_resp = requests.get(
```
Comment (rbikar) [inline ubi_manifest/app/api.py:104]: I think we should catch exceptions here  - this can raise e.g. `ConnectionError` and other exceptions as well (like TimeoutError) and then whole API call to ubi-manifest status endpoint will fail. We should try to catch most of exceptions if not all of them. 
Also I'm thinking about adding some retries as well - but I guess I'd wait for this to see how it behaves with real systems.

**Review by rbikar — CHANGES_REQUESTED** [review]

File: ubi_manifest/app/api.py (hunk 10 above)
Comment (drepelov) [reply ubi_manifest/app/api.py:104]: Okay I guess I don't understand still. I thought you do NOT want to raise the exception so the whole status endpoint would NOT fail if pulp or gitlab fails (for whatever reason). I thought we want the status endpoint to successfully return their status, even though they are failing.
You said 
> I think we should catch all exceptions raised from this call and set the status as n/a or failed.

But what is the point in setting the gitlab or pulp status to anything, if you want to raise the exception, which would prevent the endpoint to return that status but would fail with respective error and user wouldn't know where is the problem unless they check the logs from worker where it is visible what exactly was the problem.
So do we want the whole endpoint to fail or to return successfully even with some failed statuses? 
Or am I still missing something here?

Comment (rbikar) [reply ubi_manifest/app/api.py:104]: This endpoint should return 200 with all available statuses. It doesn't matter if some dependency is not OK, it's up to client what they will do with returned data.

So any check for status you do in this endpoint should be robust enough so it will never fail whole endpoint.

**Review by rbikar — APPROVED** [review]
